pkg/scheduler/plugins/gpushare/device_cache.go

```diff
@@ -0,0 +1,132 @@
+limitations under the License.
+*/
+
+package gpushare
+
+import (
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	quotav1 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+)
+
+type nodeDeviceCache struct {
+	lock sync.Mutex
+	// nodeDeviceInfos stores nodeDevice for each node
+	// and uses node name as map key.
+	nodeDeviceInfos map[string]*nodeDevice
+}
+
+type nodeDevice struct {
+	lock        sync.Mutex
+	deviceTotal map[schedulingv1alpha1.DeviceType]deviceResources
+	deviceFree  map[schedulingv1alpha1.DeviceType]deviceResources
+	deviceUsed  map[schedulingv1alpha1.DeviceType]deviceResources
+	allocateSet map[string]*framework.PodInfo
+}
+
+// deviceResources is used to present resources per device.
+// we use the minor of device as key
+// "0": {koordinator.sh/gpu-core:100, koordinator.sh/gpu-memory-ratio:100, koordinator.sh/gpu-memory: 16GB}
+// "1": {koordinator.sh/gpu-core:100, koordinator.sh/gpu-memory-ratio:100, koordinator.sh/gpu-memory: 16GB}
+type deviceResources map[int]corev1.ResourceList
+
+func newNodeDeviceCache() *nodeDeviceCache {
+	return &nodeDeviceCache{
+		nodeDeviceInfos: make(map[string]*nodeDevice),
+	}
+}
+
+func newNodeDevice() *nodeDevice {
+	return &nodeDevice{
+		deviceTotal: make(map[schedulingv1alpha1.DeviceType]deviceResources),
+		deviceFree:  make(map[schedulingv1alpha1.DeviceType]deviceResources),
+		deviceUsed:  make(map[schedulingv1alpha1.DeviceType]deviceResources),
+		allocateSet: make(map[string]*framework.PodInfo),
+	}
+}
+
+func (n *nodeDeviceCache) getNodeDevice(nodeName string) *nodeDevice {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	return n.nodeDeviceInfos[nodeName]
+}
+
+func (n *nodeDevice) resetDeviceFree(dType schedulingv1alpha1.DeviceType) {
+	for minor, usedResource := range n.deviceUsed[dType] {
+		n.deviceFree[dType][minor] = quotav1.SubtractWithNonNegativeResult(
+			n.deviceTotal[dType][minor],
+			usedResource)
+	}
+}
+
+func (n *nodeDevice) updateDeviceUsed(dType schedulingv1alpha1.DeviceType, allocation *apiext.DeviceAllocation, add bool) {
+	if add {
+		n.deviceUsed[dType][int(allocation.Minor)] = quotav1.Add(
+			n.deviceUsed[dType][int(allocation.Minor)],
+			allocation.Resources)
+	} else {
+		n.deviceUsed[dType][int(allocation.Minor)] = quotav1.SubtractWithNonNegativeResult(
+			n.deviceUsed[dType][int(allocation.Minor)],
+			allocation.Resources)
+	}
+}
+
+func (n *nodeDevice) tryAllocateGPU(podRequest corev1.ResourceList) []*apiext.DeviceAllocation {
+	var allocateResult []*apiext.DeviceAllocation
+	if isMultipleGPUPod(podRequest) {
+		gpuCore, gpuMem, gpuMemRatio := podRequest[apiext.GPUCore], podRequest[apiext.GPUMemory], podRequest[apiext.GPUMemoryRatio]
+		gpuWanted := gpuCore.Value() % 100
```

Comment:
`gpuWanted` maybe zero... if gpuCore.Value() return 100.. 

gpuWanted := gpuCore.Value() % 100 + 1

Reply:
my bad, gpuWanted shoud be `gpuCore.Value() / 100`